### PR TITLE
Make DH & EC EVP_PKEY_CTX parameter ctrls / setters more available

### LINKED ISSUE #13550: Make algorithm specific EVP_PKEY_CTX setters and getters more available
Algorithm specific EVP_PKEY_CTX setters and getters have generally only been made available when the algorithm in question is enabled.  So for example, when `no-rsa` is configured, `EVP_PKEY_CTX_set_rsa_padding()` and `EVP_PKEY_CTX_get_rsa_padding()` aren't available.

As long as only methods that are built-in in libcrypto are available, that's fine, because it reflects what implementations are available.
However, with ENGINEs and providers that may implement algorithms regardless of OpenSSL configuration, having those functions disabled becomes more questionable.

This has been changed already for DSA, because of #13529, fixed by #13530.
The rest of the key types for which we have this sort of functionality needs to be updated the same way: DH, RSA and EC_KEY

### PR DESCRIPTION
**Pull request by levitte**
DH & EC EVP_PKEY_CTX_set_ functions were only available when DH / EC was enabled ('no-dh' or 'no-ec' not configured).  However, that makes it impossible to use these functions with an engine or a provider that happens to implement DH or EC_KEY.  This change solves that problem by shuffling these functions to more appropriate places.

Partially fixes #13550

-----

This is similar to #13530

### CODE REVIEW

**Comment by t8m** [comment]
GitHub CI / buildtest (pull_request)  and GitHub CI / enable_non-default_options (pull_request)  failures look relevant.

**Comment by levitte** [comment]
That made me look more closely at all the `OPENSSL_NO_DH` and `OPENSSL_NO_EC`, and quite a few of them are not necessary any more.

**Comment by levitte** [comment]
A deep dive in the `OPENSSL_NO_DH`, `OPENSSL_NO_DSA` and `OPENSSL_NO_EC` checks, quite a number of them unnecessary or even plain wrong, and the changes became a bit... more substantial.

I believe that the transition to more accessible EVP_PKEY_CTX setters and getters is now fairly complete, at least for DH, DSA and EC.

**Comment by levitte** [comment]
CIs are happy

**Comment by t8m** [comment]
The buildtest failure seems relevant

**Comment by levitte** [comment]
Rebased, now that #13599 is merged.

**Comment by levitte** [comment]
> The buildtest failure seems relevant

I believe I fixed it...  there was an extra `extern "C" {` that I had forgotten to remove.

**Comment by levitte** [comment]
Yup, only the constant sanitizer failure remains...

File: crypto/evp/ec_support.c [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+        if ((nid = ec_curve_nist2nid_int(name)) != NID_undef)
+            return nid;
+
+#ifndef FIPS_MODULE
+        /*
+         * TODO(3.0) Figure out if we can use other names than the NIST names
+         * ("B-163", "K-163" & "P-192") in the FIPS module, or if other names
+         * are allowed as well as long as they lead to the same curve data.
+         * If only the NIST names are allowed in the FIPS module, we should
+         * move '#endif' to just before 'return NID_undef'.
```
Comment (t8m) [inline crypto/evp/ec_support.c:154]: There is no reason why non-NIST names for the NIST curves should not be allowed in the FIPS module.  You can drop this TODO.

Comment (levitte) [reply crypto/evp/ec_support.c:154]: Good, thanks

**Comment by levitte** [comment]
Rebased to fix merge conflicts

File: include/openssl/ec.h [hunk 2]
```diff
@@ -18,6 +18,54 @@
...
+# define EVP_PKEY_CTRL_GET_EC_KDF_MD            (EVP_PKEY_ALG_CTRL + 6)
+# define EVP_PKEY_CTRL_EC_KDF_OUTLEN            (EVP_PKEY_ALG_CTRL + 7)
+# define EVP_PKEY_CTRL_GET_EC_KDF_OUTLEN        (EVP_PKEY_ALG_CTRL + 8)
+# define EVP_PKEY_CTRL_EC_KDF_UKM               (EVP_PKEY_ALG_CTRL + 9)
+# define EVP_PKEY_CTRL_GET_EC_KDF_UKM           (EVP_PKEY_ALG_CTRL + 10)
+
+/* KDF types */
+# define EVP_PKEY_ECDH_KDF_NONE                      1
+# define EVP_PKEY_ECDH_KDF_X9_63                     2
+/** The old name for EVP_PKEY_ECDH_KDF_X9_63
```
Comment (FdaSilvaYY) [inline include/openssl/ec.h:63]: Style nit :  /** should be on his own line.

File: include/openssl/ec.h [hunk 3]
```diff
@@ -18,6 +18,54 @@
...
+ *  The ECDH KDF specification has been mistakingly attributed to ANSI X9.62,
```
Comment (FdaSilvaYY) [inline include/openssl/ec.h:64]: typo : mistakingly

File: crypto/evp/ec_support.c [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+                return curve_list[i].nid;
+        }
+    }
+
+    return NID_undef;
+}
+
+/* Functions to translate between common NIST curve names and NIDs */
+
+static EC_NAME2NID nist_curves[] = {
```
Comment (FdaSilvaYY) [inline crypto/evp/ec_support.c:159]: static const ?

File: include/openssl/ec.h (hunk 2 above)
Comment (levitte) [reply include/openssl/ec.h:63]: Fixed

File: include/openssl/ec.h (hunk 3 above)
Comment (levitte) [reply include/openssl/ec.h:64]: Fixed

File: crypto/evp/ec_support.c (hunk 4 above)
Comment (levitte) [reply crypto/evp/ec_support.c:159]: Fixed

**Comment by levitte** [comment]
Ping

**Review by t8m — APPROVED** [review]

**Comment by levitte** [comment]
Merged

d33ab074ef9847b67d96961f85f4ad614395d2c2 EVP_PKEY & DSA: move dsa_ctrl.c to be included only on libcrypto
c829c23b67308ad8e8ab677c78db1d5151106c3c EVP_PKEY & DH: Make DH EVP_PKEY_CTX parameter ctrls / setters more available
565b33990cc03d757f493616c040addbedfc80f8 EVP_PKEY & EC_KEY: Make EC EVP_PKEY_CTX parameter ctrls / setters more available
c2403f362efc519e473269d183adc7cceddb8c54 Drop unnecessary checks of OPENSSL_NO_DH, OPENSSL_NO_DSA and OPENSSL_NO_EC
a2e145f8dbf064b56f793fc77b3ccd79f7c67689 Add necessary checks of OPENSSL_NO_DH, OPENSSL_NO_DSA and OPENSSL_NO_EC

**Pull request closed without merging**